site/docs/skylark/macros.md

```diff
@@ -7,12 +7,131 @@ title: Macros
 
 <!-- [TOC] -->
 
-## Macro creation
+## Introduction
 
 A macro is a function called from the BUILD file that can instantiate rules.
-Macros don't give additional power, they are just used for encapsulation and
-code reuse. By the end of the [loading phase](concepts.md#evaluation-model),
-macros don't exist anymore, and Bazel sees only the set of rules they created.
+Macros are mainly used for encapsulation and code reuse of existing rules 
+and other macros. By the end of the [loading phase](concepts.md#evaluation-model), 
+macros don't exist anymore, and Bazel sees only the concrete set of instantiated rules.
+
+## Usage
+
+The typical use-case for a macro is when you want to reuse a rule. 
+
+For example, we have a genrule in a BUILD file that generates a file 
+using `//test:generator` with a `some_arg` argument hardcoded in the command:
+
+```python
+genrule(
+    name = "file",
+    outs = ["file.txt"],
+    cmd = "$(location generator) some_arg > $@",
+    tools = ["//test:generator"],
+)
+```
+
+> Tip: `$@` is a [Make variable](../be/make-variables.html#predefined_genrule_variables) 
+> that refers to the execution-time locations of the files in the `outs` attribute list. 
+> It is equivalent to `$(locations :file.txt)`.
+
+If you want to generate more files with different arguments, you may want to
+extract this code to a macro function. Let's call the macro `file_generator`, which
+has `name` and `arg` parameters. Replace the genrule with the following:
+
+```python
+load("//path:generator.bzl", "file_generator")
+
+file_generator(
+    name = "file",
+    arg = "some_arg",
+)
+
+file_generator(
+    name = "file-two",
+    arg = "some_arg_two",
+)
+
+file_generator(
+    name = "file-three",
+    arg = "some_arg_three",
+)
+```
+
+Here, we are loading the `file_generator` symbol from a `.bzl` file located in the 
+`//path` package. By putting macro function definitions in a separate `.bzl` file, 
+we can keep our BUILD files clean and declarative,  The `.bzl` file can be loaded 
+from any package in the workspace, as long as `visibility` conditions are met.
+
+> Tip: To make a `.bzl` file visible from a package, add [`exports_files(["<filename>.bzl"])`](../be/functions.html#exports_files) into the `BUILD` file in the same directory as the `.bzl` file.
+
+Finally, in `path/generator.bzl`, let's write the definition of the macro to encapsulate and 
+parameterize our original genrule definition:
+
+```python
+def file_generator(name, arg, visibility=None):
+  native.genrule(
+    name = name,
+    outs = [name + ".txt"],
+    cmd = "$(location //test:generator) %s > $@" % arg,
+    tools = ["//test:generator"],
+    visibility = visibility,
+  )
+```
+
+You can also use macros to chain rules together. This example shows a generated
+text file undergoing conversions to markdown and rich text format:
+
+```python
+def txt_md_rtf_generator(name, visibility=None):
+  # Generates a txt file
+  native.genrule(
```

Comment:
Why bother with this. How about taking the text file name as an arg to the macro. That makes the example shorter without diminishing the point you are making.

Reply:
The name is the arg to the macro. I simplified the example.